File: python_email_server/email_parser.py
import re
import email
from email.message import EmailMessage as StandardEmailMessage
from email.policy import default
from typing import Dict, List, Any, Optional, Tuple

class ParsedEmail:
    """Simple class to hold parsed email data."""
    
    def __init__(self):
        self.from_name: str = ""
        self.from_address: str = ""
        self.to: List[str] = []
        self.subject: str = ""
        self.text: str = ""
        self.html: Optional[str] = None
        self.headers: Dict[str, Any] = {}
        self.attachments: List[Dict[str, Any]] = []


def parse_email(email_data: str) -> ParsedEmail:
    """Parse an email message and extract key information."""
    parsed = ParsedEmail()
    
    try:
        # Parse the email message
        msg = email.message_from_string(email_data, policy=default)
        
        # Extract headers
        parsed.headers = dict(msg.items())
        parsed.subject = msg.get('Subject', '')
        
        # Extract sender information
        from_header = msg.get('From', '')
        parsed.from_name, parsed.from_address = _parse_address(from_header)
        
        # Extract recipient information
        to_header = msg.get('To', '')
        for addr in to_header.split(','):
            _, email_addr = _parse_address(addr.strip())
            if email_addr:
                parsed.to.append(email_addr)
        
        # Extract body content
        _extract_content(msg, parsed)
        
    except Exception as e:
        print(f"Error parsing email: {e}")
        
    return parsed


def _parse_address(address: str) -> Tuple[str, str]:
    """Parse an email address into name and actual address."""
    name = ""
    email_addr = ""
    
    # Try to match patterns like "Name <email@example.com>" or just "email@example.com"
    match = re.match(r'"?([^"<]+)"?\s*<?([^>]*)>?', address)
    if match:
        name = match.group(1).strip()
        email_addr = match.group(2).strip() or name
    else:
        email_addr = address.strip()
        
    return name, email_addr


def _extract_content(msg: StandardEmailMessage, parsed: ParsedEmail) -> None:
    """Extract text and HTML content from the email."""
    # Check if the message is multipart
    if msg.is_multipart():
        for part in msg.iter_parts():
            content_type = part.get_content_type()
            if content_type == 'text/plain':
                parsed.text = part.get_content()
            elif content_type == 'text/html':
                parsed.html = part.get_content()
    else:
        # Not multipart, just get the content
        content_type = msg.get_content_type()
        if content_type == 'text/plain':
            parsed.text = msg.get_content()
        elif content_type == 'text/html':
            parsed.html = msg.get_content()
        else:
            parsed.text = str(msg.get_content())


def extract_magic_links(content: str) -> List[str]:
    """Extract magic links from email content."""
    # Regular expression to find URLs
    url_regex = r'https?://[^\s"<>]+'
    
    # Find all URLs in the content
    urls = re.findall(url_regex, content)
    
    # Filter to likely magic links
    magic_links = []
    for url in urls:
        lower_url = url.lower()
        if any(keyword in lower_url for keyword in [
            'token=', 'verify', 'confirm', 'reset', 
            'auth', 'magic', 'login'
        ]):
            magic_links.append(url)
            
    return magic_links

File: server/email-parser.ts
import { log } from "./vite";

// Simple interface for parsed email
export interface ParsedEmail {
  from: {
    name: string;
    address: string;
  };
  to: string[];
  subject: string;
  text: string;
  html: string | null;
  headers: Record<string, any>;
  attachments: Array<{
    filename: string;
    content: Buffer;
    contentType: string;
  }>;
}

/**
 * Simple email parser to extract key information from email data
 */
export async function parseEmail(emailData: string): Promise<ParsedEmail> {
  try {
    const lines = emailData.split(/\r?\n/);
    const headers: Record<string, string> = {};
    const attachments: any[] = [];
    
    let i = 0;
    
    // Parse headers
    while (i < lines.length && lines[i] !== '') {
      const line = lines[i];
      
      if (line.match(/^\s/)) {
        // Continuation of previous header
        const lastHeader = Object.keys(headers).pop();
        if (lastHeader) {
          headers[lastHeader] += ' ' + line.trim();
        }
      } else {
        // New header
        const match = line.match(/^([^:]+):\s*(.*)$/);
        if (match) {
          const name = match[1].trim().toLowerCase();
          const value = match[2].trim();
          headers[name] = value;
        }
      }
      
      i++;
    }
    
    // Skip empty line after headers
    i++;
    
    // Find boundary if multipart
    let boundary = null;
    const contentType = headers['content-type'] || '';
    const boundaryMatch = contentType.match(/boundary="?([^";\r\n]+)"?/i);
    
    if (boundaryMatch) {
      boundary = boundaryMatch[1];
    }
    
    let text = '';
    let html = null;
    
    // Parse body
    if (boundary) {
      // Multipart message
      const parts = emailData.split('--' + boundary);
      
      for (let part of parts) {
        if (part.includes('Content-Type: text/plain')) {
          const partLines = part.split(/\r?\n\r?\n/);
          if (partLines.length > 1) {
            text = partLines.slice(1).join('\n\n');
          }
        } else if (part.includes('Content-Type: text/html')) {
          const partLines = part.split(/\r?\n\r?\n/);
          if (partLines.length > 1) {
            html = partLines.slice(1).join('\n\n');
          }
        }
      }
    } else {
      // Single part message
      text = lines.slice(i).join('\n');
    }
    
    // Extract from name and address
    let fromName = '';
    let fromAddress = '';
    
    const from = headers['from'] || '';
    const fromMatch = from.match(/"?([^"<]+)"?\s*<?([^>]*)>?/);
    
    if (fromMatch) {
      fromName = fromMatch[1].trim();
      fromAddress = fromMatch[2].trim() || fromMatch[1].trim();
    } else {
      fromAddress = from;
    }
    
    // Extract to addresses
    const to: string[] = [];
    const toHeader = headers['to'] || '';
    const toAddresses = toHeader.split(',');
    
    for (let addr of toAddresses) {
      const toMatch = addr.match(/"?([^"<]+)"?\s*<?([^>]*)>?/);
      if (toMatch && toMatch[2]) {
        to.push(toMatch[2].trim());
      } else {
        to.push(addr.trim());
      }
    }
    
    return {
      from: {
        name: fromName,
        address: fromAddress
      },
      to,
      subject: headers['subject'] || '',
      text: text.trim(),
      html: html ? html.trim() : null,
      headers,
      attachments
    };
  } catch (error) {
    log(`Error parsing email: ${error}`);
    
    // Return a minimal valid object in case of parsing errors
    return {
      from: { name: '', address: '' },
      to: [],
      subject: '',
      text: '',
      html: null,
      headers: {},
      attachments: []
    };
  }
}

/**
 * Extract magic links from email text content
 */
export function extractMagicLinks(content: string): string[] {
  const urlRegex = /(https?:\/\/[^\s"<>]+)/g;
  const matches = content.match(urlRegex) || [];
  
  // Filter to likely magic links (containing tokens, auth, verify, etc.)
  return matches.filter(url => {
    const lowercaseUrl = url.toLowerCase();
    return (
      lowercaseUrl.includes('token=') ||
      lowercaseUrl.includes('verify') ||
      lowercaseUrl.includes('confirm') ||
      lowercaseUrl.includes('reset') ||
      lowercaseUrl.includes('auth') ||
      lowercaseUrl.includes('magic') ||
      lowercaseUrl.includes('login')
    );
  });
}
